File: Fortran/Geodetics/Fukushima/HalleyMethod/testConvertEcefToGeodetic.f
C===============================================================================
C        1         2         3         4         5         6         7         8
C2345678901234567890123456789012345678901234567890123456789012345678901234567890
C===============================================================================
       PROGRAM
     2 testConvertEcefToGeodetic
C-------------------------------------------------------------------------------
C
C   MAIN PROGRAM:
C
C     testConvertEcefToGeodetic 
C 
C-------------------------------------------------------------------------------
C
C   PURPOSE:
C 
C     To test the conversion of specified ECEF retangular coordinates
C     to geodetic coordinates for a specified reference ellipsoid.
C 
C-------------------------------------------------------------------------------
C 
C   METHOD:
C 
C     [ 1 ] Uses the economic third-order Halley's method to solve the
C           general non-linear fourth-order algebraic geodetic equation
C           numerically.
C 
C     [ 2 ] Uses only  one iteration of the iterative Halley's method
C           to achieve near double precision accuracy.
C 
C     [ 3 ] Uses a technique to avoid division operations which significantly
C           accelerates the backward transformation without degrading the
C           precision.
C 
C-------------------------------------------------------------------------------
C 
C   AUTHOR(s):
C 
C     [ 1 ]  Toshio Fukushima <Toshio.Fukushima@nao.ac.jp>
C            National Astronomical Observatory of Japan (NAOJ)
C            Address:  2-21-1, Ohsawa, Mitaka, Tokyo   181-8588,  Japan
C            Phone:    +81-422-34-3613
C 
C-------------------------------------------------------------------------------
C 
C   REFERENCE(s):
C 
C      [ 1 ]  "Transformation from Cartesian to geodetic
C              coordinates accelerated by Halley's method",
C             Toshio Fukushima,
C             J.Geodesy (2006),
C             Volume 79,
C             Pages 689-693
C 
C      [ 2 ]  "Fast transform from geocentric to geodetic coordinates",
C             Toshio Fukushima,
C             Journal Of Geodesy (1999),
C             Volume 73,
C             Pages 603–610
C 
C      [ 3 ]  "Geometric Geodesy, Part A",
C             "A set of lecture notes which are an introduction to
C              ellipsoidal geometry related to geodesy.",
C              R. E. Deakin and M. N. Hunter,
C              School of Mathematical and Geospatial Sciences,
C              RMIT University,
C              Melbourne, Australia,
C              January 2013
C              www.mygeodesy.id.au/documents/Geometric%20Geodesy%20A(2013).pdf
C 
C      [ 4 ]  'Various parameterizations of "latitude" equation -
C              Cartesian to geodetic coordinates transformation',
C              Marcin Ligas,
C              Journal of Geodetic Science,
C              Pages 87 - 94,
C              2013
C 
C      [ 5 ]  "In numerical analysis, Halley's method is a root-finding
C              algorithm used for functions of one real variable with a
C              continuous second derivative.",
C             "The rate of convergence of the iterative Halley's method
C              is cubic.",
C             "There exist multidimensional versions of Halley's method.",
C             wikipedia.org/wiki/Halley's_method`
C
C-------------------------------------------------------------------------------
       IMPLICIT   NONE
C-------------------------------------------------------------------------------
       INTEGER*4  latitudeIndex
       INTEGER*4  altitudeIndex
       INTEGER*4  minimumLatitudeIndex
       INTEGER*4  maximumLatitudeIndex
       INTEGER*4  minimumAltitudeIndex
       INTEGER*4  maximumAltitudeIndex
C     - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
       INTEGER*4  STANDARD_OUTPUT
C     - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
       REAL*8     deltaGeodeticLatitudeDegrees
       REAL*8     deltaGeodeticAltitudeMeters
       REAL*8     minimumTrueGeodeticNorthLatitudeDegrees
       REAL*8     maximumTrueGeodeticNorthLatitudeDegrees
C     - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
       REAL*8     PI
       REAL*8     radiansPerDegree
       REAL*8     degreesPerRadian
       REAL*8     arcMinutesPerDegree
       REAL*8     arcSecondsPerArcMinute
       REAL*8     arcSecondsPerDegree
       REAL*8     microArcSecondsPerArcSecond
       REAL*8     microArcSecondsPerDegree
       REAL*8     microArcSecondsPerRadian
       REAL*8     nanoMetersPerMeter
       REAL*8     earthEquatorialRadiusMeters
       REAL*8     inverseEarthEllipsoidalFlatteningFactor
       REAL*8     earthEllipsoidalFlatteningFactor
C     - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
       REAL*8     earthEllipsoidalEccentricitySquared
       REAL*8     trueGeocentricEastLongitudeDegrees
       REAL*8     trueGeocentricEastLongitudeRadians
       REAL*8     trueGeodeticNorthLatitudeDegrees
       REAL*8     trueGeodeticNorthLatitudeRadians
       REAL*8     trueGeodeticAltitudeMeters
       REAL*8     xTrueRectangularMeters
       REAL*8     yTrueRectangularMeters
       REAL*8     zTrueRectangularMeters
       REAL*8     estimatedGeocentricEastLongitudeRadians
       REAL*8     estimatedGeodeticNorthLatitudeRadians
       REAL*8     estimatedGeodeticAltitudeMeters
       REAL*8     geodeticAltitudeErrorMeters
       REAL*8     geodeticAltitudeErrorNanoMeters
       REAL*8     geodeticLatitudeErrorRadians
       REAL*8     geodeticLatitudeErrorMicroArcSeconds
       REAL*8     maximumGeodeticNorthLatitudeAbsoluteErrorMicroArcSecs
       REAL*8     maximumGeodeticAltitudeAbsoluteErrorNanoMeters
C-------------------------------------------------------------------------------
       STANDARD_OUTPUT             = 6
C     - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
       PI                          = 4.0D0   * DATAN( 1.0D0 )
       radiansPerDegree            = PI      /  180.0D0
       degreesPerRadian            = 180.0D0 /  PI
C     - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
       arcMinutesPerDegree         =  60.0D0
       arcSecondsPerArcMinute      =  60.0D0
       nanoMetersPerMeter          =   1.0D9
       microArcSecondsPerArcSecond =   1.0D6
       arcSecondsPerDegree         =  arcSecondsPerArcMinute      *
     2                                arcMinutesPerDegree
       microArcSecondsPerDegree    =  microArcSecondsPerArcSecond *
     2                                arcSecondsPerDegree
       microArcSecondsPerRadian    =  microArcSecondsPerDegree    *
     3                                degreesPerRadian
C     --------------------------------------------------------------------------
C
C      The parameters:
C        [ 1 ]  Earth equatorial radius [meters]
C        [ 2 ]  Inverse Earth ellipsoidal flattening factor
C      are set to the GRS1980 System values.
C
C     --------------------------------------------------------------------------
       earthEquatorialRadiusMeters             = 6378137.0D0
       inverseEarthEllipsoidalFlatteningFactor =     298.257222101D0
C     --------------------------------------------------------------------------
       maximumGeodeticNorthLatitudeAbsoluteErrorMicroArcSecs = 0.0D0
       maximumGeodeticAltitudeAbsoluteErrorNanoMeters        = 0.0D0
C     --------------------------------------------------------------------------
       earthEllipsoidalFlatteningFactor        =
     2                           1.0D0 /
     3                           inverseEarthEllipsoidalFlatteningFactor
C     --------------------------------------------------------------------------
C
C      STRATEGY:
C
C        [ 1 ]  Define true geodetic latitude and altitude
C               values along the 45-degree East geocentric
C               longitude meridian.
C
C        [ 2 ]  Generate true rectangular values based on the
C               true geodetic latitude, true geodetic altitude
C               and constant geocentric longitude values.
C
C        [ 3 ]  Compute estimated geodetic values based on the
C               true rectangular values.
C
C        [ 4 ]  Report the differences between the defined true
C               geodetic values and the estimated geodetic
C               values.
C
C     --------------------------------------------------------------------------
       earthEllipsoidalEccentricitySquared     =
     2                 ( 2.0D0 - earthEllipsoidalFlatteningFactor ) *
     3                           earthEllipsoidalFlatteningFactor
C     --------------------------------------------------------------------------
       CALL
     2 generateTestProgramPurposeMessage(  )
C     --------------------------------------------------------------------------
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A )'
     5  )
     6  '',
     7  '',
     8  ''
C     --------------------------------------------------------------------------
C
C      Fix the true longitude at 45 degrees East for test purposes.
C
C     --------------------------------------------------------------------------
       trueGeocentricEastLongitudeDegrees      = 45.0D0
C     --------------------------------------------------------------------------
       minimumTrueGeodeticNorthLatitudeDegrees =  1.0D-9
       maximumTrueGeodeticNorthLatitudeDegrees = 89.999999999D0
C     --------------------------------------------------------------------------
       CALL
     2 generateTestProgramOutputHeader
     3   (
     4     trueGeocentricEastLongitudeDegrees
     5   )
C     --------------------------------------------------------------------------
       trueGeocentricEastLongitudeRadians =
     2                            radiansPerDegree *
     3                            trueGeocentricEastLongitudeDegrees
C     --------------------------------------------------------------------------
       deltaGeodeticLatitudeDegrees = 15.0D+0
       deltaGeodeticAltitudeMeters  =  1.0D+6
C     --------------------------------------------------------------------------
       minimumAltitudeIndex         =   0;
       maximumAltitudeIndex         =   3;
       minimumLatitudeIndex         =  -1;
       maximumLatitudeIndex         =   6;
C     --------------------------------------------------------------------------
C
C      Loop for true Latitude values.
C      (You may increase the number of test data set)
C
C     --------------------------------------------------------------------------
       DO  latitudeIndex = minimumLatitudeIndex,
     2                     maximumLatitudeIndex
C         {---------------------------------------------------------------------
           trueGeodeticNorthLatitudeDegrees =
     2                              DBLE( latitudeIndex ) *
     3                              deltaGeodeticLatitudeDegrees
C         ----------------------------------------------------------------------
C
C          Include Near-Equator/Polar Cases
C
C         ----------------------------------------------------------------------
           trueGeodeticNorthLatitudeDegrees =
     2              MIN(
     3                   maximumTrueGeodeticNorthLatitudeDegrees,
     4                   MAX(
     5                        trueGeodeticNorthLatitudeDegrees,
     6                        minimumTrueGeodeticNorthLatitudeDegrees
     7                      )
     8                 )
C         ----------------------------------------------------------------------
           trueGeodeticNorthLatitudeRadians =
     2                              radiansPerDegree *
     3                              trueGeodeticNorthLatitudeDegrees
C         ----------------------------------------------------------------------
C
C          Loop for true Altitude values.
C          (You may increase the number of test data set)
C
C         ----------------------------------------------------------------------
           DO  altitudeIndex = minimumAltitudeIndex,
     2                         maximumAltitudeIndex
C             {-----------------------------------------------------------------
               IF( altitudeIndex .EQ. 0 )
     2         THEN
C                {--------------------------------------------------------------
                   trueGeodeticAltitudeMeters = -1.0D4
C                }--------------------------------------------------------------
               ELSE
C                {--------------------------------------------------------------
                   trueGeodeticAltitudeMeters =
     2                                 DBLE( altitudeIndex )
     3                                 *
     4                                 deltaGeodeticAltitudeMeters
C                }--------------------------------------------------------------
               ENDIF
C             ------------------------------------------------------------------
C              Generate true rectagular values based on the defined true
C              geodetic values.
C             ------------------------------------------------------------------
               xTrueRectangularMeters = 0.0D0
               yTrueRectangularMeters = 0.0D0
               zTrueRectangularMeters = 0.0D0
C             - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
               CALL
     2         convertGeodeticToEcef
     3                (
     4                  earthEquatorialRadiusMeters,
     5                  earthEllipsoidalEccentricitySquared,
     6                  trueGeodeticNorthLatitudeRadians,
     7                  trueGeocentricEastLongitudeRadians,
     8                  trueGeodeticAltitudeMeters,
     9                  xTrueRectangularMeters,
     A                  yTrueRectangularMeters,
     B                  zTrueRectangularMeters
     C                )
C             ------------------------------------------------------------------
C
C              Estimate the geodetic values based on the true rectangular
C              coordinate values using Halley's third order method
C
C              (applying only for one iteration to obtain desired double
C               precision accuracy)
C
C             ------------------------------------------------------------------
               estimatedGeodeticNorthLatitudeRadians   = 0.0D0
               estimatedGeocentricEastLongitudeRadians = 0.0D0
               estimatedGeodeticAltitudeMeters         = 0.0D0
C             - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
               CALL
     2         convertEcefToGeodetic
     3           (
     4             earthEquatorialRadiusMeters,
     5             earthEllipsoidalFlatteningFactor,
     6             xTrueRectangularMeters,
     7             yTrueRectangularMeters,
     8             zTrueRectangularMeters,
     9             estimatedGeodeticNorthLatitudeRadians,
     A             estimatedGeocentricEastLongitudeRadians,
     B             estimatedGeodeticAltitudeMeters
     C           )
C             ------------------------------------------------------------------
C
C              Generate the error residuals
C
C             ------------------------------------------------------------------
               geodeticLatitudeErrorRadians =
     2                      trueGeodeticNorthLatitudeRadians -
     3                 estimatedGeodeticNorthLatitudeRadians
               geodeticLatitudeErrorMicroArcSeconds =
     2                 microArcSecondsPerRadian      *
     3                 geodeticLatitudeErrorRadians
C             - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
               geodeticAltitudeErrorMeters =
     2                 trueGeodeticAltitudeMeters -
     3                 estimatedGeodeticAltitudeMeters
               geodeticAltitudeErrorNanoMeters =
     2                 nanoMetersPerMeter          *
     3                 geodeticAltitudeErrorMeters
C             ------------------------------------------------------------------
               maximumGeodeticNorthLatitudeAbsoluteErrorMicroArcSecs
     2         =
     3         dmax1
     4          (
     5            maximumGeodeticNorthLatitudeAbsoluteErrorMicroArcSecs,
     6            dabs( geodeticLatitudeErrorMicroArcSeconds )
     7          )
C             - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
               maximumGeodeticAltitudeAbsoluteErrorNanoMeters
     2         =
     3         dmax1
     4          (
     5            maximumGeodeticAltitudeAbsoluteErrorNanoMeters,
     6            dabs( geodeticAltitudeErrorNanoMeters )
     7          )
C             ------------------------------------------------------------------
C              Generate a report.
C             ------------------------------------------------------------------
               WRITE
     2           (
     3             STANDARD_OUTPUT,
     4             '( 3X, 0PF16.10,  0PF18.5, 0PES22.10, 0PES22.10  )'
     5           )
     6           trueGeodeticNorthLatitudeDegrees,
     7           trueGeodeticAltitudeMeters,
     8           geodeticLatitudeErrorMicroArcSeconds,
     9           geodeticAltitudeErrorNanoMeters
C             ------------------------------------------------------------------
C              End of Altitude loop.
C             }-----------------------------------------------------------------
           END DO
C         ----------------------------------------------------------------------
C          End of Latitude loop.
C         }---------------------------------------------------------------------
       END DO
C     --------------------------------------------------------------------------
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A, A )'
     5  )
     6  '=============================================================',
     7  '==========================='
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A )'
     5  )
     6  '',
     7  '',
     8  ''
C     --------------------------------------------------------------------------
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A, A )'
     5  )
     6  '============================================',
     7  '============================================'
C     --------------------------------------------------------------------------
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A )'
     5  )
     6  '|',
     7  '|  MAXIMUM ABSOLUTE ERRORS OVER ONE TRIAL:',
     8  '|',
     9  '|    Maximum geodetic north latitude absolute error'
C     --------------------------------------------------------------------------
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A, 0PF14.6, A )'
     5  )
     6  '|      is:-->',
     7   maximumGeodeticNorthLatitudeAbsoluteErrorMicroArcSecs,
     8  ' [microarcseconds]'
C     --------------------------------------------------------------------------
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A )'
     5  )
     6  '|',
     7  '|    Maximum geodetic altitude absolute error'
C     --------------------------------------------------------------------------
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A, 0PF14.6, A )'
     5  )
     6  '|      is:-->',
     7  maximumGeodeticAltitudeAbsoluteErrorNanoMeters,
     8  ' [nanometers]'
C     --------------------------------------------------------------------------
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A )'
     5  )
     6 '|'
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A, A )'
     5  )
     6 '============================================',
     7 '============================================'
C     --------------------------------------------------------------------------
       WRITE
     2  (
     3    STANDARD_OUTPUT,
     4    '( A )'
     5  )
     6  '',
     7  '',
     8  ''
C     --------------------------------------------------------------------------
       STOP
C     --------------------------------------------------------------------------
       END
C===============================================================================

File: C++/Geodetics/Fukushima/HalleyMethod/testConvertEcefToGeodetic.cpp
//==============================================================================
//       1         2         3         4         5         6         7         8
//345678901234567890123456789012345678901234567890123456789012345678901234567890
//==============================================================================

#include "conversionBetweenEcefAndGeodetic.h"

//------------------------------------------------------------------------------
int
main
 (
   //-------------------
   // INPUT(s):
   //-------------------
      const int    numberCommandLineArguments,
      const char *pVectorCommandLineArguments[ ]
   //-------------------
   // OUTPUT(s):
   //           NONE
   //-------------------
 )
//==============================================================================
//
//  MAIN PROGRAM:
//    testConvertEcefToGeodetic 
//
//------------------------------------------------------------------------------
//
//  PURPOSE:
//
//    To test the conversion of specified Earth Centered Earth Fixed (ECEF)
//    retangular coordinates to geodetic coordinates for a specified
//    reference ellipsoid.
//
//------------------------------------------------------------------------------
//
//  METHOD:
//
//    [ 1 ] Uses the economic third-order Halley's method to solve the
//          general non-linear fourth-order algebraic geodetic equation
//          numerically.
//
//    [ 2 ] Uses only  one iteration of the iterative Halley's method
//          to achieve near double precision accuracy.
//
//    [ 3 ] Uses a technique to avoid division operations which significantly
//          accelerates the backward transformation without degrading the
//          precision.
//
//------------------------------------------------------------------------------
//
//  AUTHOR(s):
//
//    [ 1 ]  Toshio Fukushima <Toshio.Fukushima@nao.ac.jp>
//           National Astronomical Observatory of Japan (NAOJ)
//           Address:  2-21-1, Ohsawa, Mitaka, Tokyo   181-8588,  Japan
//           Phone:    +81-422-34-3613
//
//------------------------------------------------------------------------------
//
//  REFERENCE(s):
//
//     [ 1 ]  "Transformation from Cartesian to geodetic
//             coordinates accelerated by Halley's method",
//            Toshio Fukushima,
//            J.Geodesy (2006),
//            Volume 79,
//            Pages 689-693
//
//     [ 2 ]  "Fast transform from geocentric to geodetic coordinates",
//            Toshio Fukushima,
//            Journal Of Geodesy (1999),
//            Volume 73,
//            Pages 603–610
//
//     [ 3 ]  "Geometric Geodesy, Part A",
//            "A set of lecture notes which are an introduction to
//             ellipsoidal geometry related to geodesy.",
//             R. E. Deakin and M. N. Hunter,
//             School of Mathematical and Geospatial Sciences,
//             RMIT University,
//             Melbourne, Australia,
//             January 2013
//             www.mygeodesy.id.au/documents/Geometric%20Geodesy%20A(2013).pdf
//
//     [ 4 ]  'Various parameterizations of "latitude" equation -
//             Cartesian to geodetic coordinates transformation',
//             Marcin Ligas,
//             Journal of Geodetic Science,
//             Pages 87 - 94,
//             2013
//
//     [ 5 ]  "In numerical analysis, Halley's method is a root-finding
//             algorithm used for functions of one real variable with a
//             continuous second derivative.",
//            "The rate of convergence of the iterative Halley's method
//             is cubic.",
//            "There exist multidimensional versions of Halley's method.",
//            wikipedia.org/wiki/Halley's_method`
//
//==============================================================================
{
//------------------------------------------------------------------------------
   int     mainProgramReturnValue = 0;
//------------------------------------------------------------------------------
   const
   double  piRadians                      = 4.0 * atan( 1.0 );
   const
   double  radiansPerDegree               = piRadians / 180.0;
   const
   double  degreesPerRadian               = 180.0     / piRadians;
   const
   double  arcSecondsPerArcMinute         =  60.0;
   const
   double  arcMinutesPerDegree            =  60.0;
   const
   double  microArcSecondsPerSecond       =   1.0E6;
   const
   double  microArcSecondsPerDegree       = microArcSecondsPerSecond *
                                            arcSecondsPerArcMinute   *
                                            arcMinutesPerDegree;
   const
   double  microArcSecondsPerRadian       = degreesPerRadian         *
                                            microArcSecondsPerDegree;
   const
   double  nanoMetersPerMeter             = 1.0E9;
//------------------------------------------------------------------------------
//
// The parameters:
//   [ 1 ]  Earth equatorial radius [meters]
//   [ 2 ]  Inverse Earth ellipsoidal flattening factor
// are set to the GRS1980 System values.
//
//------------------------------------------------------------------------------
   const
   double  earthEquatorialRadiusMeters    = 6378137.0E0;
   const
   double  inverseEarthEllipsoidalFlatteningFactor
                                          =     298.257222101E0;
//------------------------------------------------------------------------------
   const
   double  earthEllipsoidalFlatteningFactor
                                          =
                                1.0E0 /
                                inverseEarthEllipsoidalFlatteningFactor;
   const
   double  earthEllipsoidalEccentricitySquared
                                          = (
                                              2.0E0 -
                                              earthEllipsoidalFlatteningFactor
                                            )
                                            *
                                              earthEllipsoidalFlatteningFactor;
//------------------------------------------------------------------------------
// Fixed the longitude at 45 degrees for test purposes.
//------------------------------------------------------------------------------
   const
   double  specifiedFixedTrueEastGeocentricLongitudeDegrees = 45.0;
//==============================================================================
//
// STRATEGY:
//
//   [ 1 ]  Define true geodetic latitude and altitude
//          values along the 45-degree East geocentric
//          longitude meridian.
//
//   [ 2 ]  Generate true rectangular values based on the
//          true geodetic latitude, true geodetic altitude
//          and constant geocentric longitude' values.
//
//   [ 3 ]  Compute estimated geodetic values based on the
//          true rectangular values
//
//   [ 4 ]  Report the differences between the defined true
//          geodetic values and the estimated geodetic
//          values.
//
//==============================================================================
   const
   double  specifiedFixedTrueEastGeocentricLongitudeRadians =
                        radiansPerDegree *
                        specifiedFixedTrueEastGeocentricLongitudeDegrees;
//------------------------------------------------------------------------------
//
// Generate a purpose message.
//
//------------------------------------------------------------------------------
   generateTestProgramPurposeMessage(  );
//------------------------------------------------------------------------------
//
// Generate Header for Output
//
//------------------------------------------------------------------------------
   generateTestProgramOutputHeader
           (
             specifiedFixedTrueEastGeocentricLongitudeDegrees
           );
//------------------------------------------------------------------------------
   const
   double  deltaLatitudeDegrees = 15.0;
   const
   int     numberLatitudeSteps  =  6;
   const
   int     numberAltitudeSteps  =  3;
//------------------------------------------------------------------------------
   const
   double  TEN_KILOMETERS_IN_METERS           =   10000.0;
   const
   double  ONE_THOUSAND_KILOMETERS_IN_METERS  = 1000000.0;
//------------------------------------------------------------------------------
   const
   double  minimumTrueGeodeticLatitudeDegrees =  0.000000001;
   const
   double  maximumTrueGeodeticLatitudeDegrees = 89.999999999;
//------------------------------------------------------------------------------
   double  maximumGeodeticNorthLatitudeAbsoluteErrorMicroArcSecs = 0.0;
   double  maximumGeodeticAltitudeAbsoluteErrorNanoMeters        = 0.0;
//------------------------------------------------------------------------------
//
// Loop over Geodetic Latitude values.
//
//------------------------------------------------------------------------------
   for(
        int latitudeIndex  = -1;
            latitudeIndex <=  numberLatitudeSteps;
            latitudeIndex  =  latitudeIndex + 1
      )
      {
       //-----------------------------------------------------------------------
          double  trueNorthGeodeticLatitudeDegrees = latitudeIndex *
                                                     deltaLatitudeDegrees;
       //-----------------------------------------------------------------------
          trueNorthGeodeticLatitudeDegrees   =
              fmin(
                    maximumTrueGeodeticLatitudeDegrees,
                    fmax(
                          trueNorthGeodeticLatitudeDegrees,
                          minimumTrueGeodeticLatitudeDegrees
                        )
                  );
       //-----------------------------------------------------------------------
          const
          double
           trueGeodeticNorthLatitudeRadians  = radiansPerDegree *
                                               trueNorthGeodeticLatitudeDegrees;
       //--------------- -------------------------------------------------------
       //
       // Loop over Geodetic Altitude values.
       //
       //-----------------------------------------------------------------------
          for(
               int altitudeIndex  =  0;
                   altitudeIndex <=  numberAltitudeSteps;
                   altitudeIndex  =  altitudeIndex + 1
             )
             {
              //----------------------------------------------------------------
                 double trueGeodeticAltitudeMeters = 0.0;
              //----------------------------------------------------------------
                 if( altitudeIndex == 0 )
                   {
                    //----------------------------------------------------------
                       trueGeodeticAltitudeMeters = -TEN_KILOMETERS_IN_METERS;
                    //----------------------------------------------------------
                   }
                 else
                   {
                    //----------------------------------------------------------
                       trueGeodeticAltitudeMeters =
                                           altitudeIndex *
                                           ONE_THOUSAND_KILOMETERS_IN_METERS;
                    //----------------------------------------------------------
                   };
              //----------------------------------------------------------------
              // Convert the true geodetic coordinates into Earth Centered
              // Earth Fixed (ECEF) noninertial coordinates.
              //----------------------------------------------------------------
                 double xTrueEcefMeters = NAN;
                 double yTrueEcefMeters = NAN;
                 double zTrueEcefMeters = NAN;
              //- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
                 ( void )
                 convertGeodeticToEcef
                        (
                         //----------------
                         // INPUT(s):
                         //----------------
                            earthEquatorialRadiusMeters,
                            earthEllipsoidalEccentricitySquared,
                            trueGeodeticNorthLatitudeRadians,
                            specifiedFixedTrueEastGeocentricLongitudeRadians,
                            trueGeodeticAltitudeMeters,
                         //----------------
                         // OUTPUT(s):
                         //----------------
                            xTrueEcefMeters,
                            yTrueEcefMeters,
                            zTrueEcefMeters
                        );
              //----------------------------------------------------------------
              // Use Halley's third order iterative method
              // (applying only one iteration) to approximate the solution of
              // the fourth order nonlinear algebraic geodetic equations.
              //----------------------------------------------------------------
                 double estimatedGeodeticNorthLatitudeRadians   = 0.0;
                 double estimatedGeocentricEastLongitudeRadians = 0.0;
                 double estimatedGeodeticAltitudeMeters         = 0.0;
              //- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
                 ECEF_TO_GEODETIC_CONVERSION_STATUS
                   functionReturnedStatus =
                   convertEcefToGeodetic
                          (
                            //-----------------
                            // INPUT(s):
                            //-----------------
                               earthEquatorialRadiusMeters,
                               earthEllipsoidalFlatteningFactor,
                               xTrueEcefMeters,
                               yTrueEcefMeters,
                               zTrueEcefMeters,
                            //-----------------
                            // OUTPUT(s):
                            //-----------------
                               estimatedGeodeticNorthLatitudeRadians,
                               estimatedGeocentricEastLongitudeRadians,
                               estimatedGeodeticAltitudeMeters
                          );
              //----------------------------------------------------------------
                 if( functionReturnedStatus == SUCCESSFUL_CONVERSION )
                   {
                    //----------------------------------------------------------
                    // Geocentric to Geodetic conversion was successful.
                    //- - - - - - - - - - - - - - - - - - - - - - - - - - - - -
                    //
                    // Compute the residuals (errors):
                    //
                    //----------------------------------------------------------
                       const
                        double
                        deltaGeodeticLatitudeRadians =
                                     trueGeodeticNorthLatitudeRadians -
                                     estimatedGeodeticNorthLatitudeRadians;
                        const
                         double
                         deltaGeodeticLatitudeMicroArcSec = 
                                      microArcSecondsPerRadian  *
                                      deltaGeodeticLatitudeRadians;
                    //- - - - - - - - - - - - - - - - - - - - - - - - - -
                        const
                         double
                         deltaGeodeticAltitudeMeters =
                                      trueGeodeticAltitudeMeters -
                                      estimatedGeodeticAltitudeMeters;
                        const
                         double
                         deltaGeodeticAltitudeNanoMeters  =
                                      nanoMetersPerMeter *
                                      deltaGeodeticAltitudeMeters;
                    //----------------------------------------------------------
                        maximumGeodeticNorthLatitudeAbsoluteErrorMicroArcSecs =
                        std::max
                        (
                          maximumGeodeticNorthLatitudeAbsoluteErrorMicroArcSecs,
                          abs( deltaGeodeticLatitudeMicroArcSec )
                        );
                    //- - - - - - - - - - - - - - - - - - - - - - - - - - - - -
                        maximumGeodeticAltitudeAbsoluteErrorNanoMeters =
                        std::max
                        (
                          maximumGeodeticAltitudeAbsoluteErrorNanoMeters,
                          abs( deltaGeodeticAltitudeNanoMeters )
                        );
                    //----------------------------------------------------------
                    // Output results.
                    //----------------------------------------------------------
                       fprintf
                         (       
                           stdout,
                           "%+19.10f %+18.5f  %+19.10e  %+18.10e\n",
                           trueNorthGeodeticLatitudeDegrees,
                           trueGeodeticAltitudeMeters,
                           deltaGeodeticLatitudeMicroArcSec,
                           deltaGeodeticAltitudeNanoMeters
                         );
                    //----------------------------------------------------------
                   }
                 else
                   {
                    //----------------------------------------------------------
                    // Geocentric to Geodetic conversion has failed.
                    //- - - - - - - - - - - - - - - - - - - - - - - - - - - - -
                    // Generate an error message.
                    //----------------------------------------------------------
                       if(
                           functionReturnedStatus
                           ==
                           INVALID_ELLIPSOIDAL_FLATTENING
                         )
                         {
                          //----------------------------------------------------
                          // Encountered invalid Earth ellipsoidal
                          // flattening value.
                          //- - - - - - - - - - - - - - - - - - - - - - - - - -
                          // Generate an error message.
                          //----------------------------------------------------
                             fprintf
                              (
                                stdout,
                                "\n\n\n"
                                "%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n"
                                "%s\n%s\n%s\n"
                                "%s%+20e\n"
                                "%s\n%s\n"
                                "\n\n\n",
                                "---------------------------------------------",
                                "|",
                                "| ERROR:",
                                "|",
                                "|   Encountered invalid Earth ellipsoidal",
                                "|   flattening value.",       
                                "|",
                                "|   The Earth ellipsoidal flattening values",
                                "|   should be in the interval:",
                                "|     [ 0.0, 1.0 ).",
                                "|",
                                "|   The Earth ellipsoidal flattening value",
                                "|   was:   ",
                                earthEllipsoidalFlatteningFactor,
                                "|",
                                "---------------------------------------------"
                              );
                          //----------------------------------------------------
                         }
                       else
                       if
                         (
                           functionReturnedStatus
                           ==
                           INVALID_EQUATORIAL_RADIUS
                         )
                         {
                          //----------------------------------------------------
                          // Encountered invalid Earth equatorial radius.
                          //- - - - - - - - - - - - - - - - - - - - - - - - - -
                          // Generate an error message.
                          //----------------------------------------------------
                             fprintf
                              (
                                stdout,
                                "\n\n\n"
                                "%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n"
                                "%s\n%s\n"
                                "%s%+20e%s\n"
                                "%s\n%s\n"
                                "\n\n\n",
                                "---------------------------------------------",
                                "|",
                                "| ERROR:",
                                "|",
                                "|   Encountered invalid Earth",       
                                "|   equatorial radius value.",
                                "|",
                                "|   The Earth equatorial radius",
                                "|   value should be strictly positive.",
                                "|",
                                "|   The Earth equatorial radius value",
                                "|   was:   ",
                                earthEquatorialRadiusMeters,
                                " [meters]",                    
                                "|",
                                "---------------------------------------------"
                              );
                          //----------------------------------------------------
                         }
                       else
                         {
                          //----------------------------------------------------
                          // Encountered undetermined Geodetic
                          // conversion error.
                          //- - - - - - - - - - - - - - - - - - - - - - - - - -
                          // Generate an error message.
                          //----------------------------------------------------
                             fprintf
                              (
                                stdout,
                                "\n\n\n"
                                "%s\n%s\n%s\n%s\n%s\n%s\n%s\n%s\n"
                                "\n\n\n",
                                "---------------------------------------------",
                                "|",
                                "| ERROR:",
                                "|",
                                "|   Encountered undetermined Geodetic",
                                "|   conversion error.",
                                "|",
                                "---------------------------------------------"
                              );
                                //----------------------------------------------
                         };
                          //----------------------------------------------------
                   };
              //----------------------------------------------------------------
             };
       //-----------------------------------------------------------------------
      };
//------------------------------------------------------------------------------
// Finish the output report.
//------------------------------------------------------------------------------
   fprintf
    (
      stdout,
      "%s%s"
      "\n\n\n",
      "============================================",
      "============================================"
    );
//------------------------------------------------------------------------------
   fprintf
    (
      stdout,
        "\n\n\n"
        "%s%s\n"
        "%s\n%s\n%s\n%s\n"
        "%s%+14.6e%s\n"
        "%s\n%s\n"
        "%s%+14.6e%s\n"
        "%s\n"
        "%s%s\n"
        "\n\n\n",
      "============================================",
      "============================================",
      "|",
      "|  MAXIMUM ABSOLUTE ERRORS OVER ONE TRIAL:",
      "|",
      "|    Maximum geodetic north latitude absolute error",
      "|      is:-->",
       maximumGeodeticNorthLatitudeAbsoluteErrorMicroArcSecs,
      " [microarcseconds]",
      "|",
      "|    Maximum geodetic altitude absolute error",
      "|      is:-->",
      maximumGeodeticAltitudeAbsoluteErrorNanoMeters,
      " [nanometers]",
      "|",
      "============================================",
      "============================================"
    );
//------------------------------------------------------------------------------
   return( mainProgramReturnValue );
//------------------------------------------------------------------------------
}
//==============================================================================
